### ISSUE TITLE
Gemfile require option allows you to require any file

### ISSUE THREAD

**Issue by ngan** [rubygems/rubygems#6453]
### Describe the problem as clearly as you can

Within the `Gemfile`, you can do:

```ruby
gem 'foo', require: ["foo", "foo/some/other/file/in/gem"]
```

This is useful if you want to automatically `require` some other entrypoint for the gem.

Today, I just realized that this simply passes the value to `Kernel.require`, which means you can require _any_ file in the filesystem.

```ruby
gem 'foo', require: ["/some/ruby/file/in/fs", "../../../some/other/file/relative/to/path"]
```
It seems intuitive if the files that you require in the `gem` directive are scoped to that gem. I'm not sure if this is a vulnerability since it's code that you own/written, but I think a little lockdown here would be good. Ideally, whatever is passed through to `require: [...]` is validated and prepended to the gem's path. The validation would simply not allow absolute paths in `require` or a relative path that attempts to traverse outside the gem's path.

**Comment by deivid-rodriguez** [comment]
I agree `require:` should be scoped to path inside the target gem 👍.

**Comment by ngan** [comment]
I tried to implement this and got it working. However, it breaks gems like `rails`, which has a "meta-gem" that doesn't contain anything. The `lib/rails.rb` actually resides in the `railtie` gem. Maybe we just ensure that you can only require a file that lives in one of the paths in `$LOAD_PATH`?

**Comment by indirect** [comment]
I think I agree with @ngan and we will probably need to allow at most limited anything inside $LOAD_PATH for backwards compatibility reasons.

**Comment by deivid-rodriguez** [comment]
Nice catch! Yeah I think that makes sense too. Alternatively, we could forbid absolute paths or non absolute paths starting with "." (making the underlying require relative to the CWD).

**Comment by bensheldon** [comment]
Alternative viewpoint: I prefer the way it currently works because it makes it easy to patch in a custom setup script immediately before the entrypoint for a gem is loaded:

```ruby
gem "pg", require: [File.expand_path('config/pg_preload.rb', __dir__), "pg"]
```

Having this flexibility makes it easier to rely solely on `Bundler.require` and the Gemfile for (unfortunate) load order dependencies.
